src/mam4xx/mo_sethet.hpp

```diff
@@ -174,6 +176,308 @@ void find_ktop(
+      gas_washout(team, kk, xkgm, xliq(kk), // in
+                  xhen_h2o2, tfld, delz,    // in
+                  xeqca, xca,
+                  xgas2);                   // inout
+      gas_washout(team, kk, xkgm, xliq(kk), // in
+                  xhen_so2, tfld, delz,     // in
+                  xeqca, xca,
+                  xgas3); // inout
+    }
+    //-----------------------------------------------------------------
+    //       ... calculate the lifetime of washout (second)
+    //             after all layers washout
+    //             the concentration of hno3 is reduced
+    //             then the lifetime xtt is calculated by
+    //
+    //                  xtt = (xhno3(ini) - xgas1(new))/(dt*xhno3(ini))
+    //                  where dt = passing time (s) in vertical
+    //                             path below the cloud
+    //                        dt = dz(cm)/um(cm/s)
+    //-----------------------------------------------------------------
+    xdtm = delz(kk) / xum; // the traveling time in each dz
+
+    xxx2 = (xh2o2(kk) - xgas2(kk));
+    if (xxx2 != 0.0) { // if no washout lifetime = 1.e29
+      yh2o2 = xh2o2(kk) / xxx2 * xdtm;
+    } else {
+      yh2o2 = large_value_lifetime;
+    }
+    tmp_hetrates[1](kk) = haero::max(1.0 / yh2o2, 0.0) * stay;
+
+    xxx3 = (xso2(kk) - xgas3(kk));
+    if (xxx3 != 0.0) { // if no washout lifetime = 1.e29
+      yso2 = xso2(kk) / xxx3 * xdtm;
+    } else {
+      yso2 = large_value_lifetime;
+    }
+    tmp_hetrates[2](kk) = haero::max(1.0 / yso2, 0.0) * stay;
+  }
+
+  //-----------------------------------------------------------------
+  //       ... part 2, in-cloud solve for low henry constant
+  //                   hno3 and h2o2 have both in and under cloud
+  //-----------------------------------------------------------------
+  for (int kk = ktop; kk < pver; kk++) {
+    for (int mm = 0; mm < gas_pcnst; mm++) {
+      if (rain(kk) <= 0.0) {
+        het_rates[mm](kk) = 0.0;
+      }
+    }
+
+    for (int kk = ktop; kk < pver; kk++) {
+
+      work1 = avo2 * xliq(kk);
+      work2 = const0 * tfld(kk);
+
+      if (h2o2_ndx >= 0) {
+        calc_het_rates(satf_h2o2, rain(kk), xhen_h2o2(kk), // in
+                       tmp_hetrates[1](kk), work1, work2,  // in
+                       het_rates[h2o2_ndx](kk));           // out
+      }
+
+      // if ( prog_modal_aero .and.
+      if (so2_ndx >= 0 && h2o2_ndx >= 0) {
+        het_rates[so2_ndx](kk) = het_rates[h2o2_ndx](kk);
+      } else if (so2_ndx >= 0) {
+        calc_het_rates(satf_so2, rain(kk), xhen_so2(kk),  // in
+                       tmp_hetrates[2](kk), work1, work2, // in
+                       het_rates[so2_ndx](kk));           // out
+      }
+
+      if (h2so4_ndx >= 0) {
+        calc_het_rates(satf_hno3, rain(kk), xhen_hno3(kk), // in
+                       tmp_hetrates[0](kk), work1, work2,  // in
+                       het_rates[h2so4_ndx](kk));          // out
```

Comment:
Similar to the input index questions, these indices of `tmp_hetrates` are hard-coded in the fortran. I've deprecated them here. I don't believe they match the `_ndx` that is being checked but sanity checking that with y'all?